### ISSUE TITLE
When is the 'restorestate' event fired?

### ISSUE THREAD

**Issue by majido** [majido/scroll-restoration-proposal#3]
From Jonas Sicking:

When we designed the pushState() API we ran into problems with a
similar event which caused some significant problems and required a
last-minute redesign. The resulting design for pushState() is that the
'popstate' event does not fire when the user does a session-history
traversal that navigates between Documents. There simply was no good
time to fire such an event.

If we fired 'popstate' before the 'load' event, then the <script
>
which added the event listener might not have loaded yet. If we fired
'popstate' after the 'load' event then the page didn't know what state
to render the page in until the 'load' event fired, which results in
crappy UX.

Instead we made the current state available through history.state, and
then 'popstate' only fires for sessionhistory transitions within the
same Document, i.e. when history.state changes.

So when does 'restorestate' fire? Does it fire before or after 'load'?
Keep in mind that browsers start restoring the scroll state long
before 'load' but doesn't load all <script>s until 'load' has been
fired.

**Comment by majido** [comment]
Thanks for raising this issue and providing the context.

I don't think we limit the 'restorestate' to only fire for same-document transitions. It will reduce its value significantly :(.

I believe the earliest time it can be fired is when 'DOMContentReady' is fired which is after sync script tags are parsed but before 'load'.  This still can mean delaying the scroll restoration until then which is certainly not a great for UX. I believe right now we are able to restore the scroll position event before the first frame is rendered. 

I think this is serious problem with the event base approach. The history API solution does not have the same problem.

**Comment by RByers** [comment]
This is a really good point.  If this is likely to become a real problem then it would be a good reason to prefer a state-associated design (like originally proposed) instead of an event-based one.  @smaug---  WDYT?

**Comment by sicking** [comment]
Yeah, I think this means that the event based approach doesn't work.

**Comment by stuartpb** [comment]
I think you could let content scripts set a listener that can only happen after all the scripts load, and if they need to handle things _before_ content scripts load, they provide it with a mechanism akin to Service Worker that stores a copy of the script that needs to fire before / as the page starts loading.

**Comment by stuartpb** [comment]
But yeah, that thing I just described would be the introduction of a way crazy mechanism that is primarily there to solve problems app developers can mostly solve by just doing synchronous operations with state values made available on first load, and by setting states to control the way those saved values are applied.
